### PR TITLE
chore(artifacts): setup future usage of xxhash artifact digest

### PR DESCRIPTION
**Pull request by amusipatla-wandb**
Description
-----------

- Addresses [WB-35725](https://coreweave.atlassian.net/browse/WB-35725)
- Addresses [WB-35726](https://coreweave.atlassian.net/browse/WB-35726)

This PR sets up the artifact class to use xxhash128 as a digest algorithm. In particular, it adds support for this as a digest algorithm for local uploads via `add_file` and `add_dir`. it also adds support for this type of digest when caching artifacts during download.

With this PR, the old behavior of using MD5 is maintained (manifest is formatted the same) and we default to MD5 for the digest algorithm since this PR doesn't include a fallback to MD5 if necessary when uploading. The PR following will change this to default to xxhash always, and will add a fallback to MD5 if necessary, and alter the tests that upload artifacts to match.

Some implementation details: we keep track of the digest algorithm in each manifest entry (if its XXH128) for backwards compatibility for some edge cases. We also choose to default `job` artifacts to MD5 since they have separate artifact builder logic.

[Feature design doc](https://app.notion.com/p/wandbai/Migration-strategies-for-upgrading-artifact-hashing-algorithm-374e2f5c7ef3804285b9fa2b5a662659?source=copy_link)

- [x] I updated CHANGELOG.unreleased.md, or it's not applicable

Testing
-------
- Added unit tests around caching and making sure downloads use the correct cache paths
- Added unit test to make sure incremental artifacts are using MD5 if that is what the parent used
- Added system tests for add_file/add_dir, using the correct algorithm for digest calculations, and building the manifest as expected.

[WB-35725]: https://coreweave.atlassian.net/browse/WB-35725?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[WB-35726]: https://coreweave.atlassian.net/browse/WB-35726?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: wandb/sdk/artifacts/artifact_digest_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+"""Artifact digest algorithm."""
+
+from __future__ import annotations
+
+from enum import Enum
+
+
+class ArtifactDigestAlgorithm(str, Enum):
+    MANIFEST_MD5 = "MANIFEST_MD5"
+    MANIFEST_XXH128 = "MANIFEST_XXH128"
```
Comment (tonyyli-wandb) [inline wandb/sdk/artifacts/artifact_digest_algorithm.py]: **suggestion:** This can be codegen'ed since you added it to the GraphQL schema.  That said, there _are_ good reasons to define a hand-crafted enum outside of generated code -- one of them would be to simplify the enum names (while keeping the actual values unchanged):

```suggestion
class DigestAlgorithm(str, Enum):
    MD5 = "MANIFEST_MD5"
    XXH128 = "MANIFEST_XXH128"
```

...something like this would probably make internal code much much less verbose and more readable

File: wandb/sdk/artifacts/artifact.py [hunk 2]
```diff
@@ -177,6 +183,7 @@ def __init__(
 
         # Internal.
         self._service_api: ServiceApi | None = None
+        self._digest_algorithm: str | None = None
```
Comment (tonyyli-wandb) [inline wandb/sdk/artifacts/artifact.py]: **suggestion/question:** should this just be the enum type?  I.e. probably no reason to leave it as an unconstrained `str`

```suggestion
        self._digest_algorithm: DigestAlgorithm | None = None
```

**Review by tonyyli-wandb — COMMENTED** [review]
Have to come back to this, but couple small comments in the meantime

File: tests/system_tests/test_artifacts/test_wandb_artifacts.py [hunk 3]
```diff
@@ -87,37 +88,85 @@ def test_unsized_manifest_entry():
     assert "No such file" in str(e.value)
 
 
-def test_add_one_file(artifact):
+@mark.parametrize(
+    "digest_algorithm",
+    [ArtifactDigestAlgorithm.MANIFEST_XXH128, ArtifactDigestAlgorithm.MANIFEST_MD5],
+)
```
Comment (tonyyli-wandb) [inline tests/system_tests/test_artifacts/test_wandb_artifacts.py:94]: **praise,suggestion(non-blocking):** nice! Also quick tip -- assuming we want this test to keep up with future additions to `ArtifactDigestAlgorithm`, Python enums are natively iterable.  Eg (similar suggestion for all the tests below):



```suggestion
@mark.parametrize(
    "digest_algorithm",
    ArtifactDigestAlgorithm,
)
```

File: wandb/sdk/artifacts/artifact.py [hunk 4]
```diff
@@ -1060,9 +1068,32 @@ def digest(self) -> str:
...
+
+        Once computed, the value is cached for the lifetime of the artifact.
+        """
+        if self._digest_algorithm is not None:
+            return self._digest_algorithm
+
+        # Default to MD5 unless any entry explicitly uses xxHash.
+        # TODO: default to xxh128 once we have the fallback to md5 implemented in saver.go
+        if any(
+            entry.digest_algorithm == ArtifactDigestAlgorithm.MANIFEST_XXH128
```
Comment (tonyyli-wandb) [inline wandb/sdk/artifacts/artifact.py]: **suggestion:** compare enums by identity instead of equality (`is` instead of `==`), here and everywhere else in the PR:


```suggestion
            entry.digest_algorithm is ArtifactDigestAlgorithm.MANIFEST_XXH128
```


Context: this is general Python convention/best practice, and TBH I can't think of any real exceptions to this right now.

There are reasons for this besides style -- eg off the top of my head:

- it's more performant
  - `is` compares memory addresses (by definition), and enum members are singletons (by definition)
  - I'd say this is usually marginal, but it actually might matter a bit here, since we're potentially looping over _all_ manifest entries
- it's safer
  - This is rarer and doesn't matter here, but the enum types might override the `__eq__()` method for any reason (doesn't even have to return a `bool`)
  - ...which changes behavior for `==` comparisons.  OTOH there's literally no way to override behavior for `is` comparisons

File: wandb/sdk/artifacts/artifact.py [hunk 5]
```diff
@@ -1778,9 +1812,22 @@ def _add_local_file(
...
         entry = ArtifactManifestEntry(
             path=name,
-            digest=digest or md5_file_b64(upload_path),
+            digest=digest
+            or (
+                xxh128_file_b64(upload_path)
+                if use_xxh128
+                else md5_file_b64(upload_path)
+            ),
+            digest_algorithm=entry_digest_algorithm,
```
Comment (tonyyli-wandb) [inline wandb/sdk/artifacts/artifact.py]: **suggestion:** I'd strongly consider moving most or all of this logic (i.e. instantiation logic + default values) into the `ArtifactManifest` and/or `ArtifactManifestEntry` classes themselves.

Otherwise we have to repeat this code (or something like it) anywhere we instantiate a manifest / manifest entry

Will try to follow up with separate comment(s) in the manifest/manifest entry changes after looking through those

File: wandb/sdk/artifacts/artifact_manifest.py [hunk 6]
```diff
@@ -48,7 +48,7 @@ def to_manifest_json(self) -> dict[str, Any]:
         raise NotImplementedError
 
     @abstractmethod
-    def digest(self) -> HexMD5:
+    def digest(self, digest_algorithm: str | None = None) -> HexDigest:
```
Comment (tonyyli-wandb) [inline wandb/sdk/artifacts/artifact_manifest.py]: **question:** do we need to require `digest_algorithm` as an argument here?

Gut feeling is this is unnecessarily quite risky, since it means a bug in calling code (anywhere else) could accidentally pass in a wrong/incompatible value for the `digest_algorithm` here

IIUC so far, seems like we'd either want to 
- infer the `digest_algorithm` under the hood, from looking at the manifests' contents (ie manifest entries)
- set the `digest_algorithm` on instantiating an ArtifactManifest (only if possible/appropriate)

File: wandb/sdk/artifacts/artifact_manifest_entry.py [hunk 7]
```diff
@@ -93,7 +95,8 @@ class ArtifactManifestEntry(ArtifactsBase):
 
     path: LogicalPath
 
-    digest: Union[B64MD5, ETag, URIStr, FilePathStr]
+    digest: Union[B64Digest, ETag, URIStr, FilePathStr]
+    digest_algorithm: Optional[ArtifactDigestAlgorithm] = None
```
Comment (tonyyli-wandb) [inline wandb/sdk/artifacts/artifact_manifest_entry.py]: **suggestion:** I think I see why this is optional (when it's MD5, the JSONified entry has to omit it), but I'd strongly consider instead: 
- making the field itself non-optional.  This way "MD5" is explicitly signaled by eg `DigestAlgorithm.MD5`, instead of implied by `None`
- a default value on "normal" instantiation of a fresh/new `ArtifactManifestEntry` (`MD5` now, but eventually `XXH`, IIUC)
- logic to set the correct default (MD5) _specifically when loading from JSONified data_.  Need to look around to remind myself where this is controlled from
- logic to exclude this field from JSON serialization when it's MD5

(Apologies, it occurs to me this may be too vague to be helpful -- def happy to follow up offline to be clear)

File: wandb/sdk/artifacts/artifact_digest_algorithm.py (hunk 1 above)
Comment (amusipatla-wandb) [reply wandb/sdk/artifacts/artifact_digest_algorithm.py]: I initially defined this myself because they weren't autogenerated since I didn't actually use the field in any query, but I'm changing that now so I'll switch to the codegen'ed version!

File: wandb/sdk/artifacts/artifact_manifest.py (hunk 6 above)
Comment (amusipatla-wandb) [reply wandb/sdk/artifacts/artifact_manifest.py]: yeah thanks for the pushback- I do think there's a cleaner way to set the digest_algorithm when creating the manifest now by fetching the digest algorithm and setting it accordingly rather than inferring from the types

File: wandb/sdk/artifacts/artifact_manifest_entry.py (hunk 7 above)
Comment (amusipatla-wandb) [reply wandb/sdk/artifacts/artifact_manifest_entry.py]: thanks for bringing this up because now that I'm thinking about this more, I'm not convinced we even need this at the manifest entry level now that I'm planning on fetching the digest algorithm at the artifact level and not inferring from existing entries. what we fetch there should apply to every non-reference entry regardless. so I might end up ripping this out completely, will follow up here

File: wandb/sdk/artifacts/artifact.py (hunk 5 above)
Comment (amusipatla-wandb) [reply wandb/sdk/artifacts/artifact.py]: just to clarify my understanding: are you suggesting that we should try to calculate the digest in the entry itself?I think now that I'm not keeping digest_algorithm on the entry itself this would get a little more difficult + would require extra refactoring since we are calculating the digests in some of the calling functions as well

File: wandb/sdk/artifacts/artifact_manifest_entry.py (hunk 7 above)
Comment (amusipatla-wandb) [reply wandb/sdk/artifacts/artifact_manifest_entry.py]: yes okay I moved it off of the entry and tried to just use the artifact's digest algorithm when necessary- thanks again for calling my attention to this!

File: wandb/sdk/artifacts/artifact.py (hunk 5 above)
Comment (tonyyli-wandb) [reply wandb/sdk/artifacts/artifact.py]: Ah ok.  If we're removing the `digest_algorithm` from the ManifestEntry itself, I think my original suggestion becomes either irrelevant or at least much narrower.  Can take another look once ready (lmk if ready now)

(and/or let me know if you'd prefer a quick sync)

File: wandb/sdk/artifacts/artifact.py [hunk 8]
```diff
@@ -531,7 +542,8 @@ def new_draft(self) -> Artifact:
         artifact._metadata = self.metadata
         artifact._digest_algorithm = self.digest_algorithm
         artifact._manifest = ArtifactManifest.from_manifest_json(
-            self.manifest.to_manifest_json()
+            self.manifest.to_manifest_json(),
+            self.digest_algorithm,
```
Comment (pingleiwandb) [inline wandb/sdk/artifacts/artifact.py:550]: Do we still need to pass the artifact level digest algorithm to `ArtifactManifest.from_manifest_json` when we have switched to per entry digest algo using the `alg` field in extra? Or we just need it there so we can add it for every new entry.

File: wandb/sdk/artifacts/artifact.py [hunk 9]
```diff
@@ -1804,12 +1821,26 @@ def _add_local_file(
                 os.chmod(staging_path, stat.S_IRUSR)
                 upload_path = staging_path
 
+        use_xxh128 = self.digest_algorithm is ArtifactDigestAlgorithm.MANIFEST_XXH128
```
Comment (pingleiwandb) [inline wandb/sdk/artifacts/artifact.py]: nit: I think we can have a private method/property for this check

File: wandb/sdk/artifacts/artifact.py [hunk 10]
```diff
@@ -1804,12 +1821,26 @@ def _add_local_file(
...
+            if use_xxh128
+            else {}
+        )
+
         entry = ArtifactManifestEntry(
             path=name,
-            digest=digest or md5_file_b64(upload_path),
+            digest=digest
+            or (
+                xxh128_file_b64(upload_path)
```
Comment (pingleiwandb) [inline wandb/sdk/artifacts/artifact.py]: nit: similar to previous one, we have `xxh128_file_b64(local_path) if use_xxh128 else md5_file_b64(local_path)` in previous diff L1528. We can extract these out as a private method.

File: wandb/sdk/artifacts/artifact.py [hunk 11]
```diff
@@ -2282,8 +2313,17 @@ def verify(self, root: str | None = None) -> None:
 
         ref_count = 0
         for entry in self.manifest.entries.values():
+            entry_digest_algorithm = entry.digest_algorithm()
             if entry.ref is None:
-                if md5_file_b64(validate_fspath(root, entry.path)) != entry.digest:
+                if entry_digest_algorithm is ArtifactDigestAlgorithm.MANIFEST_XXH128:
+                    file_digest = xxh128_file_b64(validate_fspath(root, entry.path))
+                elif entry_digest_algorithm is ArtifactDigestAlgorithm.MANIFEST_MD5:
```
Comment (pingleiwandb) [inline wandb/sdk/artifacts/artifact.py]: So we always set the digest algo to md5 when decoding the manifest json when `extra` is empty? Yeah, saw it in the diff of `artifact_manifest_entry.py`

File: wandb/sdk/artifacts/artifact_file_cache.py [hunk 12]
```diff
@@ -54,17 +55,28 @@ def __init__(self, cache_dir: StrPath) -> None:
...
+        algorithm: ArtifactDigestAlgorithm | None = None,
     ) -> tuple[FilePathStr, bool, Opener]:
         # Check if we're using vs skipping the cache
         if self._override_cache_path is not None:
             skip_cache = True
             path = Path(self._override_cache_path)
         else:
             skip_cache = False
-            hex_md5 = b64_to_hex_id(b64_md5)
-            path = self._obj_dir / "md5" / hex_md5[:2] / hex_md5[2:]
```
Comment (pingleiwandb) [inline wandb/sdk/artifacts/artifact_file_cache.py:67]: TIL we actually put the algo in the cache path ... I guess we predicted changing the algo  :bufo-boba:

**Review by pingleiwandb — APPROVED** [review]
Mostly nit on helper methods for some 1-3 liners. 

I am not sure if there is a more concrete name for the PR title...

What this PR does is hashing, save algo in manifest entry in python. The actual upload and md5 fallback in go core, which is another PR.

File: wandb/sdk/artifacts/artifact.py (hunk 8 above)
Comment (amusipatla-wandb) [reply wandb/sdk/artifacts/artifact.py:550]: I think since the entries can be mixed on the manifest and its no longer completely in sync with the artifact digest algorithm, this is probably safer? we just use this digest algorithm to calculate the artifact/manifest digest

File: wandb/sdk/artifacts/artifact_manifests/artifact_manifest_v1.py [hunk 13]
```diff
@@ -28,15 +29,26 @@ class ArtifactManifestV1(ArtifactManifest):
...
+    digest_algorithm: Annotated[
+        ArtifactDigestAlgorithm, Field(exclude=True, repr=False)
+    ]
+
     @classmethod
-    def from_manifest_json(cls, manifest_json: dict[str, Any]) -> ArtifactManifestV1:
+    def from_manifest_json(
+        cls,
+        manifest_json: dict[str, Any],
+        digest_algorithm: ArtifactDigestAlgorithm | None = None,
```
Comment (jacobromero) [inline wandb/sdk/artifacts/artifact_manifests/artifact_manifest_v1.py]: could we just change this to be `digest_algorithm: ArtifactDigestAlgorithm  = ArtifactDigestAlgorithm.MANIFEST_MD5`?

File: wandb/sdk/artifacts/_internal_artifact.py [hunk 14]
```diff
@@ -46,9 +47,23 @@ def __init__(
...
         )
         self._type = type
+
+        if self._type == "job":
+            # Match go-core JobBuilder / ArtifactBuilder, which always uses MD5.
+            digest_algorithm_override = ArtifactDigestAlgorithm.MANIFEST_MD5
+        elif digest_algorithm is not None:
+            digest_algorithm_override = digest_algorithm
+        else:
+            digest_algorithm_override = None
```
Comment (jacobromero) [inline wandb/sdk/artifacts/_internal_artifact.py]: you can remove the other `elif/else` if you do

```
if self._type == "job":
    disgest_algorithm = ArtifactDigestAlgorithm.MANIFEST_MD5

if digest_algorithm is not None:
...
```

File: wandb/sdk/artifacts/artifact.py [hunk 15]
```diff
@@ -1084,6 +1098,16 @@ def digest_algorithm(self) -> ArtifactDigestAlgorithm:
         """The digest algorithm used to compute the artifact's digest."""
         return self._digest_algorithm
 
+    def _digest_algorithm_is_xxh128(self) -> bool:
+        """Check if the artifact digest algorithm is xxh128."""
+        return self.digest_algorithm is ArtifactDigestAlgorithm.MANIFEST_XXH128
```
Comment (jacobromero) [inline wandb/sdk/artifacts/artifact.py]: nit: I don't think this is really useful in helping readability.

File: wandb/sdk/artifacts/artifact.py [hunk 16]
```diff
@@ -1804,12 +1829,20 @@ def _add_local_file(
                 os.chmod(staging_path, stat.S_IRUSR)
                 upload_path = staging_path
 
+        # Tag only non-MD5 entries; untagged entries are interpreted as MD5.
+        extra = (
+            {DIGEST_ALGORITHM_EXTRA_KEY: DIGEST_ALGORITHM_TO_STR[self.digest_algorithm]}
+            if self._digest_algorithm_is_xxh128()
+            else {}
+        )
```
Comment (jacobromero) [inline wandb/sdk/artifacts/artifact.py:1855]: any reason to not always include the digest algorithm?

File: wandb/sdk/artifacts/artifact.py [hunk 17]
```diff
@@ -2282,8 +2315,17 @@ def verify(self, root: str | None = None) -> None:
...
             if entry.ref is None:
-                if md5_file_b64(validate_fspath(root, entry.path)) != entry.digest:
+                if entry_digest_algorithm is ArtifactDigestAlgorithm.MANIFEST_XXH128:
+                    file_digest = xxh128_file_b64(validate_fspath(root, entry.path))
+                elif entry_digest_algorithm is ArtifactDigestAlgorithm.MANIFEST_MD5:
+                    file_digest = md5_file_b64(validate_fspath(root, entry.path))
+                else:
+                    raise ValueError(
+                        f"Invalid digest algorithm: {entry_digest_algorithm}"
+                    )
```
Comment (jacobromero) [inline wandb/sdk/artifacts/artifact.py]: I don't think this case is reachable since `entry.digest_algorithm()` always returns either `ArtifactDigestAlgorithm.MANIFEST_MD5` or `ArtifactDigestAlgorithm.MANIFEST_XXH128`, and defaults to md5 if the digest_algorithm is unrecognized, or the key is not a string value.

File: wandb/sdk/artifacts/artifact_manifest_entry.py [hunk 18]
```diff
@@ -43,6 +45,18 @@
 _WB_ARTIFACT_SCHEME: Final[str] = "wandb-artifact"
 
 
+# Consts for tracking digest algorithm in an entry's extra field.
+# Only non-MD5 entries are tagged; untagged entries are interpreted as MD5.
+DIGEST_ALGORITHM_EXTRA_KEY: Final[str] = "alg"
+DIGEST_ALGORITHM_TO_STR: Final[Dict[ArtifactDigestAlgorithm, str]] = {
+    ArtifactDigestAlgorithm.MANIFEST_MD5: "MD5",
+    ArtifactDigestAlgorithm.MANIFEST_XXH128: "XXH128",
+}
```
Comment (jacobromero) [inline wandb/sdk/artifacts/artifact_manifest_entry.py:54]: we have a few places in this PR which map `ArtifactDigestAlgorithm` enum to a string. They are all similar but very slightly different.
This makes maintenance (adding/remove digests) a pain. They aren't likely to change very often, but this could be error prone in the future.

Is there anyway we can reduce the number of times we need to do this?

File: wandb/sdk/internal/internal_api.py [hunk 19]
```diff
@@ -3236,6 +3238,7 @@ def create_artifact(
         distributed_id: str | None = None,
         is_user_created: bool | None = False,
         history_step: int | None = None,
+        digest_algorithm: ArtifactDigestAlgorithm | None = None,
```
Comment (jacobromero) [inline wandb/sdk/internal/internal_api.py]: same here, this could be `digest_algorithm: ArtifactDigestAlgorithm = ArtifactDigestAlgorithm.MANIFEST_MD5`

File: tests/system_tests/test_artifacts/test_wandb_artifacts.py
Comment (jacobromero) [inline tests/system_tests/test_artifacts/test_wandb_artifacts.py:1]: I think you missed renaming `check_md5_obj_path` to `check_digest_obj_path` in the test `test_cache_cleanup_allows_upload`.

Its marked a xfail which is why its not showing up in the test failures on this PR 🙂

**Comment by jacobromero** [comment]
it would also be nice if you could add a rationale for the change (any docs, etc.) that motivated this change in the description for future reference please 🙂

File: wandb/sdk/artifacts/artifact.py (hunk 16 above)
Comment (amusipatla-wandb) [reply wandb/sdk/artifacts/artifact.py:1855]: it's mainly a small measure to save space in the manifest file (that's also the reason for not storing MANIFEST_XXH128 and instead using "xxh128"). we can't do it the opposite way and only tag MD5 entries since existing manifests use MD5 by default and don't have this extra field, and since we have to accommodate for that anyway

File: tests/system_tests/test_artifacts/test_wandb_artifacts.py
Comment (amusipatla-wandb) [reply tests/system_tests/test_artifacts/test_wandb_artifacts.py:1]: oof, thanks for the catch!

File: wandb/sdk/artifacts/artifact.py (hunk 17 above)
Comment (amusipatla-wandb) [reply wandb/sdk/artifacts/artifact.py]: yeah it won't ever get here- mainly added this as a failsafe in case we change the entry logic in the future but probably overengineering. just going to move the MD5 logic into the else case and get rid of the elif

File: wandb/sdk/artifacts/artifact_manifest_entry.py (hunk 18 above)
Comment (amusipatla-wandb) [reply wandb/sdk/artifacts/artifact_manifest_entry.py:54]: yeah this is very fair pushback and something I'm also not really a fan of. I indicated this in an earlier comment but the reason why we are mapping from `MANIFEST_{XXH128/MD5}` to `XXH128` and `MD5` here is for space saving in the manifest file, since right now this is a limiting factor to how many files can be in an artifact.

I think the other place we do this is when determining the download cache directory (and later upload cache directory in a follow up PR I think). it doesn't really matter there, it's mostly for cleanliness and backwards compatibility since the original cache path we use has `md5`. I'm happy to reuse the mapping from this file (so we don't have to have a separate map) or to just use `MANIFEST_{XXH128/MD5}` as the directory name, we'd just have cache misses to begin with.

File: wandb/sdk/artifacts/artifact_manifests/artifact_manifest_v1.py [hunk 20]
```diff
@@ -28,15 +29,26 @@ class ArtifactManifestV1(ArtifactManifest):
...
 
         policy_name = data.storage_policy
         policy_cfg = data.storage_policy_config
         policy = StoragePolicy.lookup_by_name(policy_name).from_config(policy_cfg)
         return cls(
-            manifest_version=data.version, entries=data.contents, storage_policy=policy
+            manifest_version=data.version,
+            entries=data.contents,
+            storage_policy=policy,
+            digest_algorithm=digest_algorithm or ArtifactDigestAlgorithm.MANIFEST_MD5,
```
Comment (jacobromero) [inline wandb/sdk/artifacts/artifact_manifests/artifact_manifest_v1.py]: I don't think this `or ArtifactDigestAlgorithm.MANIFEST_MD5` should be needed with the change on line 40

**Review by jacobromero — APPROVED** [review]
